File: Problems/Snakes and Ladders/Golang/game/game.go
package game

import (
	"snakeladders/board"
	"snakeladders/dice"
	"snakeladders/jumper"
	"snakeladders/player"
	"time"
)

type Game struct {
	Jumper  []jumper.Jumper
	Board   *board.Board
	Dice    *dice.Dice
	Players []player.Player
}

var gameIntance *Game
var queue []player.Player

func newGame(jumper []jumper.Jumper, boardSize int, players []player.Player, sides uint) *Game {

	gameIntance = &Game{
		Jumper:  jumper,
		Board:   board.NewBoard(boardSize),
		Dice:    dice.NewDice(sides),
		Players: players,
	}

	queue = make([]player.Player, 0)
	queue = append(queue, players...)
	return gameIntance
}

func GetGameInstance(jumper []jumper.Jumper, boardSize int, players []player.Player, sides uint) *Game {
	if gameIntance == nil {
		gameIntance = newGame(jumper, boardSize, players, sides)
	}
	return gameIntance
}

func (g *Game) StartGame() {

	for {
		currentPlayer := queue[0]

		// Dice roll and player movement logic
		roll := g.Dice.Roll()

		println("Current Player: ", currentPlayer.GetName(), " rolled a ", roll)
		if g.Board.IsValidMove(currentPlayer.GetCurrentPosition(), int(roll)) {
			newPos := currentPlayer.GetCurrentPosition() + int(roll)
			// check for snake or ladder
			for _, jumpVal := range g.Jumper {
				if jumpVal.GetStartPosition() == newPos {
					newPos = jumpVal.GetEndPosition()
					println("Jumping from ", jumpVal.GetStartPosition(), " to ", jumpVal.GetEndPosition())
					break
				}
			}
			// check win
			if g.Board.CheckWinner(newPos) {
				println("Winner Winner chicken dinner!! player = ", currentPlayer.GetName())
				return
			}
			currentPlayer.SetCurrentPosition(newPos)
		}
		println("Current Player: ", currentPlayer.GetName(), " Position: ", currentPlayer.GetCurrentPosition())
		// Move the current player to the back of the queue

		queue = append(queue, currentPlayer)
		queue = queue[1:]
		time.Sleep(10 * time.Millisecond) // Sleep for 500 milliseconds to simulate turn-based play

	}
}

File: Problems/Snakes and Ladders/Java/game/Game.java
package game;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

import board.Board;
import dice.Dice;
import jumper.Jumper;
import player.Player;

public class Game {
    private static volatile Game instance;
    private Jumper[] jumpers;
    private Board board;
    private Dice dice;
    private Player[] players;
    private Queue<Player> playerQueue;

    private Game(Jumper[] jumpers, int boardSize, Player[] players, int sides) {
        this.board = new Board(boardSize);
        this.dice = new Dice(sides);
        this.jumpers = jumpers;
        this.players = players;
        this.playerQueue = new LinkedList<>(Arrays.asList(players));
    }

    public static Game getGameInstance(Jumper[] jumpers, int boardSize, Player[] players, int sides) {
        if (instance == null) {
            synchronized (Game.class) {
                if (instance == null) {
                    instance = new Game(jumpers, boardSize, players, sides);
                }
            }
        }
        return instance;
    }

    public void startGame(){
        while(true) {
            Player currentPlayer = playerQueue.poll();

            int roll = dice.roll();
            System.out.println("Player " + currentPlayer.getName() + " rolled a " + roll);

            if (board.isValidMove(currentPlayer.getCurrentPosition(), roll)) {
                int newPosition = currentPlayer.getCurrentPosition() + roll;

                for (Jumper jumper : jumpers) {
                    if (jumper.getStartPosition() == newPosition) {
                        newPosition = jumper.getEndPosition();
                        System.out.println("Player " + currentPlayer.getName() + " jumped from " + jumper.getStartPosition() + " to " + jumper.getEndPosition());
                        break;
                    }
                }

                if (board.checkWinner(newPosition)) {
                    System.out.println("Player " + currentPlayer.getName() + " wins!");
                    return;
                }

                currentPlayer.setCurrentPosition(newPosition);
            } 

            System.out.println("Current Player : " + currentPlayer.getName() + " at position " + currentPlayer.getCurrentPosition());
            playerQueue.offer(currentPlayer);
        }
    }

    public void getPlayers() {
        for (Player player : players) {
            System.out.println("Player: " + player.getName() + ", Position: " + player.getCurrentPosition());
        }
    }
}
